### ISSUE TITLE
IMPORTANT (CovPassCheck): Add a 'Booster'-Info to the "Certificate valid"-Screen...

### ISSUE THREAD

**Issue by Jo-Achim** [Digitaler-Impfnachweis/covpass-android#139]
## Avoid duplicates
* [X] This enhancement request has not already been raised before, but discussed; e.g. [Information über Booster bei CovPassCheck](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/128#issue-1071071159) or [CovPass shows the date of a booster vaccination as "Vollständig seit dd.mm.yyyy" and not the date of the last required vaccination](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/94#issue-1038469976) or ...
* [X] This problem exists under Android (and iOS, I think).

## Current Implementation
If vaccination certificates are scanned with **CovPassCheck** (**v1.13.3**), the current situation is as follows:
A scan of a vaccination certificate (e.g. BionTech **2 of 3**) delivers "Certificate valid".
A scan of a vaccination certificate (e.g. BionTech **3 of 3**) from the same person also provides "certificate valid".
That is correct so far.

## Suggested Enhancement
Against the background of the new regulation, which has already been partially introduced, that 'boosted' people no longer need an additional corona test, this 'boosted' information must be displayed as additional information in CovPassCheck (and CWA).

According to current information, the above-mentioned regulation - i.e. no additional tests for 'boosted' / people with a booster vaccination - should be decided in the coming week for all of Germany.

## Expected Benefits
This is the only way to determine at a glance whether an additional corona test is actually necessary or not during an access control, for example. 

E.g.:

![Screenshot_20211212-132710_21](https://user-images.githubusercontent.com/21124877/145720031-39d62c8b-f931-43d9-8397-216120361e48.jpg)

"**Booster #_1_**" is an idea to count boosters when more than one booster (i.e. a second, third, ...) booster is needed. 

## Additional context
When making the solution, it must certainly be taken into account that further booster vaccinations will be necessary for the future and of course all vaccine combinations must be taken into account when determining the 'booster'.

**Comment by vaubaehn** [comment]
Vote +1 that this is important.

It may also be integrated/combined together with https://github.com/Digitaler-Impfnachweis/covpass-android/issues/107 .
Serial scanning of certificates will be necessary to identify booster certificates (2/2) for J&J jabbed people (1/1): Both the J&J vaccination DCC and the Booster DCC need to be scanned to recognize the 2/2 as a booster.
As many people are affected, implementation of the issue here together with #107 should be done asap.

**Comment by Ein-Tim** [comment]
This is, in the very end, a duplicate of https://github.com/Digitaler-Impfnachweis/covpass-ios/issues/73. But since this issue only underlines how important this change would be, I'm fine with leaving it open.

**Comment by Ein-Tim** [comment]
@oliver-steinbrecher Any ETA? We're at 1.13 now, so this could take till beginning of February, I assume?

**Comment by oliver-steinbrecher** [comment]
Yes

**Comment by Jo-Achim** [comment]
@oliver-steinbrecher,

is it possible to change the prioritization? Or are planned changes up to v.1.17 more important?

(I know that all discussions about the relief for boosted people are a bit open in view of Omicron - according to today's reports, a decision should be made before Christmas. 
On the other hand, such reliefs could promote vaccination readiness, relieve test laboratories, ...!?)

**Comment by oliver-steinbrecher** [comment]
We’re aware of the topic’s importance. Currently I can’t confirm anything else.

**Comment by Jo-Achim** [comment]
It can currently be assumed that today (01/07/2022), at least for the gastronomy in Germany, **“2G plus” (2G +)** will be decided and will therefore come into force in the next few days.
**“2G plus” (2G +)** means, that _only_ **fully vaccinated persons plus a negative corona test** _or only_ **fully vaccinated persons plus a booster vaccination** have access to the gastronomy (and possibly other areas).

So we _need_ - especially for the controls - on the one hand a practicable handling to check the "**complete vaccination incl. test**" _and_ on the other hand the "**complete vaccination incl. booster vaccination**". This seems to me _to be a must_ and ideally should _be available at the same time_ as the regulation comes into force.

In relation to the title here: **Add a 'Booster' info to the "Certificate valid" screen ...** to CovPassCheck as soon as possible!

**Comment by Jo-Achim** [comment]
> **“2G plus” (2G +)** means, that _only_ **fully vaccinated persons plus a negative corona test** _or only_ **fully vaccinated persons plus a booster vaccination** have access to the gastronomy (and possibly other areas).
> 

And already decided!

**Comment by KaypieCode** [comment]
> > **“2G plus” (2G +)** means, that _only_ **fully vaccinated persons plus a negative corona test** _or only_ **fully vaccinated persons plus a booster vaccination** have access to the gastronomy (and possibly other areas).
> > 
> 
> And already decided!
> 
> See e.g.: [Bund und Länder einigen sich auf 2G plus für Gastronomie – deutschlandweit](https://www.spiegel.de/politik/deutschland/coronavirus-bund-und-laender-einigen-sich-auf-2g-plus-in-der-gastronomie-deutschlandweit-a-353e62bb-75e2-4668-b5ad-9261b77d9f9e)
> > In Deutschland ist der Zutritt zu Restaurants, Kneipen, Bars und Cafés für doppelt Geimpfte und Genesene bald nur noch mit einem tagesaktuellen negativen Schnelltest möglich. Darauf haben sich Bund und Länder in einer gemeinsamen Videoschalte verständigt.
> > Die sogenannte 2G-plus-Regel gilt demnach bundesweit und unabhängig von der lokalen Inzidenz. Ausgenommen von der Testpflicht sind geboosterte Personen ab dem Tag ihrer Auffrischungsimpfung. Wer weder Impf- noch Genesenennachweis erbringen kann, bleibt der Zutritt in gastronomische Betriebe gänzlich verwehrt.

@oliver-steinbrecher

**Comment by no-go** [comment]
The EU has explicitly ensured that the QR code can also be presented in paper form without a cell phone. As of today, I am therefore lost in NRW, as the CovPassCheck app cannot distinguish a Booser from an ordinary vaccination at entrance. So I have to get tested despite booster or use again the 40 years old WHO vaccination certificate.

**Comment by no-go** [comment]
Is there anyone with write access to make a review on that pull request? https://github.com/Digitaler-Impfnachweis/covpass-android/pull/156

**Comment by sleeksorrow** [comment]
FWIW: There are news articles out there where the German Federal Ministry of Health (Bundesministerium für Gesundheit, BMG) states that CovPassCheck App will not be changed to address this issue, mainly for data privacy reasons.
https://netzpolitik.org/2022/booster-nachweis-covpasscheck-app-nutzlos-bei-einigen-corona-regeln/

No idea if this ministry is capable to decide over an open source project, tho.

**Comment by Ein-Tim** [comment]
@sleeksorrow 

As the app is published by the RKI (see https://digitaler-impfnachweis-app.de/impressum), which is, according to https://www.rki.de/DE/Content/Institut/institut_node.html, "Das Robert Koch-Institut (RKI) ist ein Bundesinstitut im Geschäftsbereich des Bundesministeriums für Gesundheit."

So, in the end, I think: Yes, they can decide.

**Comment by oliver-steinbrecher** [comment]
> FWIW: There are news articles out there where the German Federal Ministry of Health (Bundesministerium für Gesundheit, BMG) states that CovPassCheck App will not be changed to address this issue, mainly for data privacy reasons. https://netzpolitik.org/2022/booster-nachweis-covpasscheck-app-nutzlos-bei-einigen-corona-regeln/
> 
> No idea if this ministry is capable to decide over an open source project, tho.

This article does not reflect the reality. As already said improvements are planed for release 1.17 . Further improvements related to booster detection are planned for release 1.18.

**Comment by sleeksorrow** [comment]
Thank you very much for the swift clarification.

**Comment by HappyLuke** [comment]
> FWIW: There are news articles out there where the German Federal Ministry of Health (Bundesministerium für Gesundheit, BMG) states that CovPassCheck App will not be changed to address this issue, mainly for data privacy reasons. https://netzpolitik.org/2022/booster-nachweis-covpasscheck-app-nutzlos-bei-einigen-corona-regeln/
> 
> No idea if this ministry is capable to decide over an open source project, tho.

I guess this is simple missinterpretation. If I understood correctly @oliver-steinbrecher they are implementing an enhanced solution to check 3g/2g/2g+ and whatever solution. The required enhancement stated in the article proposes to show the number/date of vaccination. This ministry is stating that this would not be able to implement. Checking for specific rules does not explicitly reveal more information, but is only checking for a logic in the background, so this is a different case and must be validated by data privacy department. (I assume this has already been done but not yet published).

Though I understand the CWA and the CovPass projects should be open source projects. In regards to that it is very helpful and supporting that people can raise questions and issues via GitHub. Anyhow I cannot understand why the developers do not share more information and actually discuss the requirement in more details. Making applications open source can only enhance the user experience efficiently if all stakeholders incl. the users are part of the full application life cycle. This includes the business analysis and requirement refinement process.
For this specific, very crucial topic I cannot see any further discussions. In worst case this might result in a feature which is not accepted and used by the user. So it would be very helpful for all of us if you can include people here into the discussions.
Thank you!

**Comment by Kamik423** [comment]
Just some more thoughts:

1. In most parts of Germany J&J + one booster are enough to make one count as boosted, however Bavaria has there own rules where a third vaccination is required. This has to be addressed as well.
2. Due to format limitations 2G + Booster requires two scanning passes. It should also indicate then if it scanned a test certificate or a vaccination one since most people don't put their test certificate into the CovPassApp as a EUDCC is very hard to get for a test.

In my opinion (as discussed in https://github.com/Digitaler-Impfnachweis/covpass-ios/issues/96) it would be best to provide the CovPassCheckApp operator with a menu something like this:

```
Valid checks:
[ ] Test
[ ] Vaccination
[X] Booster
[X] Vaccination + Test
```

As it is the most transparent, flexible and (in my opinion) intuitive interface for this.

**Comment by Jo-Achim** [comment]
@Kamik423,
thank you for your hints.
However, care should be taken to make it as easy as possible for the CovPassCheckApp operator. In particular, if he has to check a large number of certificates plus one ID document each. Finally, the acceptance of the measures also depends on their practical feasibility.

**Comment by al-Iskandar** [comment]
https://play.google.com/store/apps/details?id=ch.admin.bag.covidcertificate.verifier since many weeks...

**Comment by boecko** [comment]
Just wait till the wave through. So you don't have to implement that feature.
*Seriously*
* There are old people, who don't take their smartphone to every place.
* So I've printed their QRCode on a plastic card with a Zebra Card Printer.
* The were denied to enter a swimming pool, because no one could verify their Booster-Status

Am I supposed to print **"3 of 3"** on the card, so "Otto-Normal"-people accept the state of vaccination, which is in the QRCode?

**Comment by oliver-steinbrecher** [comment]
I want to let you know that development of 1.17 has been completed and the further processing is still on track.

**Comment by no-go** [comment]
@boecko that's how I see it, too.There are very official-looking EU certificates. Laminated, they look very good. This looks so official that often not even the QR code is scanned.

**Comment by no-go** [comment]
Version 1.17? There is still a big delay between f-droid (Version 1.13.2, jan 8 2022) and g-play (1.15.0, dec 21 2021) but okay :)

**Comment by gfrewqpoiu** [comment]
> @oliver-steinbrecher Version 1.17? There is still a big delay between f-droid (Version 1.13.2, jan 8 2022) and g-play (1.15.0, dec 21 2021) but okay :)

The F-Droid Version is not maintained by this team. See [here](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/141#issuecomment-994905249) for details.

**Comment by boecko** [comment]
> @boecko that's how I see it, too.There are very official-looking EU certificates. Laminated, they look very good. This looks so official that often not even the QR code is scanned.

I'm using the iOS-Screenshot from the CovpassApp. Blue background, name and then I removed the star and "Weitere Zertifikate". That looks very official, too.

**Comment by Jo-Achim** [comment]
@Kamik423 wrote: 
> In my opinion (as discussed in [Digitaler-Impfnachweis/covpass-ios#96](https://github.com/Digitaler-Impfnachweis/covpass-ios/issues/96)) it would be best to provide the CovPassCheckApp operator with a menu something like this:
> 
> ```
> Valid checks:
> [ ] Test
> [ ] Vaccination
> [X] Booster
> [X] Vaccination + Test
> ```
> 
> As it is the most transparent, flexible and (in my opinion) intuitive interface for this.

I answered: 
>> @Kamik423, thank you for your hints. However, care should be taken to make it as easy as possible for the CovPassCheckApp operator. In particular, if he has to check a large number of certificates plus one ID document each. Finally, the acceptance of the measures also depends on their practical feasibility.

I downloaded Switzerland's 'CovPassCheck app' today; see: [COVID Certificate Check](https://play.google.com/store/apps/details?id=ch.admin.bag.covidcertificate.verifier). 

In addition to the fact that the app also got along with 2G+ (booster) in a short test, @Kamik423's approach, which corresponds somewhat to the [Swiss model](https://github.com/admin-ch), is perhaps better than I initially rated it.
Why? Contrary to my idea that the 'door-keeper' should have to do as little as possible manually, the idea can also be reversed. 

The 'door keeper' usually knows what standard the guests have to have in order to be let in. As a result, it can also make sense for the examiner to set the required standard, e.g. "2G+" (once at the start of the examination), and all further scan results ultimately only show whether the certificate shown / scanned corresponds to the required standard.

Of course, it can be a bit impractical if several certificates / vaccination statuses authorize entry. But that could also be solved, if necessary, by multiple selection of the "test mode", which is currently not implemented in the current version from Switzerland but includes @Kamik423's suggestion.

So sorry @Kamik423, your suggestion is more workable than I assumed.

**Comment by vaubaehn** [comment]
> Of course, it can be a bit impractical if several certificates / vaccination statuses authorize entry. But that could also be solved, if necessary, by multiple selection of the "test mode", which is currently not implemented in the current version from Switzerland but includes @Kamik423's suggestion.

Or it's possible by implementing a serial/multiple scanning mode, as suggested in November 2021 via #107 . That feature was also implemented in Austria's GreenCheck by around November/December 2021.

**Comment by timokoenig** [comment]
The new scan mode for multiple certificates will be added in 1.17, see https://github.com/Digitaler-Impfnachweis/covpass-android/discussions/172

**Comment by vaubaehn** [comment]
@timokoenig Thanks for the hint. Heureka!

**Comment by Kamik423** [comment]
@Jo-Achim wrote:
> The 'door keeper' usually knows what standard the guests have to have in order to be let in. As a result, it can also make sense for the examiner to set the required standard, e.g. "2G+" (once at the start of the examination), and all further scan results ultimately only show whether the certificate shown / scanned corresponds to the required standard.

My words exactly :)

The best way would be probably to have presets for the configurations like "3G", "2G", "2G+ and Booster" and so on.

**Comment by oliver-steinbrecher** [comment]
> @Jo-Achim wrote:
> 
> > The 'door keeper' usually knows what standard the guests have to have in order to be let in. As a result, it can also make sense for the examiner to set the required standard, e.g. "2G+" (once at the start of the examination), and all further scan results ultimately only show whether the certificate shown / scanned corresponds to the required standard.
> 
> My words exactly :)
> 
> The best way would be probably to have presets for the configurations like "3G", "2G", "2G+ and Booster" and so on.

That’s exactly what you’ll get

**Comment by HappyLuke** [comment]
> The new scan mode for multiple certificates will be added in 1.17, see #172

Sorry, but you won't get a like for this. I can't believe scanning multiple qr codes is the best usability which can be offered by CovPass. This is exactly what I have elaborated on. The implementation is far away from the requirement described in this post. Remembering me of this cartoon https://images.app.goo.gl/RoHo44xXAWoVLxzR9

I even more like the new feature of the CWA App were 2g/3g/2g+ is shown next to the qr code. Although this is as well some kind of a workaround.

**Comment by oliver-steinbrecher** [comment]
Multiple scans are only needed in rare case - please check the regulations.

**Comment by Kamik423** [comment]
This would require multiple certificates (like multiple vaccination doses and/or recoveries) in the same QR code which is not currently permitted by the [EU Digital COVID Certificate format specification](https://ec.europa.eu/health/system/files/2021-06/covid-certificate_json_specification_en_0.pdf) (however it would be quite trivial to implement on a technical level). So there is no way to change this currently in any app. Only one vaccination dose, test, or recovery per certificate.

**Comment by Jo-Achim** [comment]
Now it may be that I misunderstand something: At least from the digital vaccination certificates, a booster can be read based on vaccination **3**/3. 
In cases where a vaccination certificate plus a test certificate has to be checked, it's will hardly be able to avoid several scans. Otherwise, a new certificate / new QR code would have to be 'created' somehow from two or more certificates.

**Comment by Kamik423** [comment]
Yes exactly. But in some cases it would be nice to have a combined certificate for vaccination + test. Currently in this case (2G+) two scans will be required currently.

**Comment by treysis** [comment]
1.17 in early February? They're joking, right?

**Comment by vaubaehn** [comment]
@treysis 
> 1.17 in early February? They're joking, right?

I assume the devs/companies are not to blame here. From SAP I know, that they're usually rather aware of necessary features (with regard to the dynamic evolution of the pandemic), and that they are negotiating with RKI implementation of new features, also taking community requests into account. And I strongly suppose the same is true for IBM.
I think the bottleneck is BMG/RKI, and I speculate that they were quite busy with themselves (also due to change in government last November/December) and due to delays of evaluation of finding (according to media, contracts have been extended just recently).

My feeling is, if needs of changes/new features are rather obvious and we don't hear so much from the devs, then BMG/RKI are the right addressees, because something is stuck there.

**Comment by treysis** [comment]
I know. My astonishment was targeted at RKI. I can't believe they're still so behind the reality. Though actually I am not surprised.

**Comment by vaubaehn** [comment]
@oliver-steinbrecher 
> Multiple scans are only needed in rare case - please check the regulations.

I guess the cases are not that rare currently: For 2G+ you won't need two DCCs, when you're boostered. As of today, around 50% of the population is boostered - and when around 75% are vaccinated, this means around 1/3 of the vaccinated don't have a booster yet. In absolute numbers that would mean roughly 27.000.000 people, that would need two DCCs to scan when entering a 2G+ venue.

So, I'm very happy about the new features (presets, serial scanning) to come.
Thanks guys!

**Comment by vaubaehn** [comment]
@Kamik423 
> This would require multiple certificates (like multiple vaccination doses and/or recoveries) in the same QR code which is not currently permitted by the [EU Digital COVID Certificate format specification](https://ec.europa.eu/health/system/files/2021-06/covid-certificate_json_specification_en_0.pdf) (however it would be quite trivial to implement on a technical level). So there is no way to change this currently in any app. Only one vaccination dose, test, or recovery per certificate.

Some additional information:
- It's not possible for a wallet app to "create" a new single DCC that contains information of multiple certificates without breaking the digital signature, as each certificate holds a signature of the hash of its certificate's contents.
- You might then say, it would be possible to "chain" two complete DCCs into one QR code, but: each single DCC already results in >700 bytes (compressed!), and I guess you followed the media/reports, that many people have problems in scanning them reliably (because of data density). If you chained two DCCs into one QR code, the QR code would need to represent data of at least 1.4KB, which likely will make scanning these QR codes very difficult to impossible in many (most?) cases.
- Integrating different information regarding immunization status (e.g., multiple vaccinations, or vaccination + recovery) into one single DCC (thus into one QR code) was rejected by the EU legal department, because - you guess it - data privacy reasons, already in early stage of development of the DCC spec. Here you will find a link that leads you to a discussion with one person in charge in the DCC development: https://github.com/Digitaler-Impfnachweis/covpass-android/issues/107#issuecomment-986045958

**Comment by treysis** [comment]
> For 2G+ you won't need two DCCs, when you're boostered.

Afaik that's not always the case and depends on state law. Bavaria still requires negative tests when entering some venues, even when boostered.

**Comment by vaubaehn** [comment]
@treysis right, I forgot about this... it's even more complex...

**Comment by vaubaehn** [comment]
@HappyLuke 
> > The new scan mode for multiple certificates will be added in 1.17, see #172
> 
> Sorry, but you won't get a like for this. I can't believe scanning multiple qr codes is the best usability which can be offered by CovPass. This is exactly what I have elaborated on. The implementation is far away from the requirement described in this post. Remembering me of this cartoon https://images.app.goo.gl/RoHo44xXAWoVLxzR9
> 
> I even more like the new feature of the CWA App were 2g/3g/2g+ is shown next to the qr code. Although this is as well some kind of a workaround.
> 
> But maybe I'm misunderstanding the new feature let's see.

I think the planned implementation is the best you can achieve with regard to a balance of usability and security (avoid abuse of fraud certificates).
If I understand the planned implementation right,
- gatekeeper will chose a preset in CovPassCheck and select the admission rule according to legislation (e.g., 2G+...)
- visitor will show the QR code from wallet app (CWA, CovPass)
- gatekeeper scans the code. In case, a second QR code is needed - and only in this case - the gatekeeper will be informed to scan a 2nd QR code and of which type. User will switch to that 2nd QR code - and in current CWA this will be just a single fingertip (afair, this will also come to CovPass)
- gatekeeper will get a signal for "pass" or "invalid"

That's it.
So, it will ease the things significantly for visitors and for gatekeepers: clear information, still best possible data privacy, no discussions anymore about the booster status....

**Comment by Kamik423** [comment]
@vaubaehn thanks for the interesting discussion. I meant the thing said in the last link: Altering the schema to support this would be very simple. The format could easily support both a test and a vaccination for example. This leave the signing issue; the app cannot do it locally. This could be solved by making a request to an EU server that can "fuse" to certificates for the same user; creating a new signed certificate after validating the ones it gets. This combined vaccine+test QR code could then be shown in big screen phones.

While I think this could improve user experience it is probably deemed as being too complex to implement.

<details>
  <summary>I think part of the issue of codes scanning badly is the specific QR code reader implemented in the CovPass apps.</summary> It seems to have some issues; especially when it is printed on a warped piece of paper such as the printout you get at a pharmacy. I have scanned far more complex barcodes with the iOS built-in QR code reader. As far as I am aware it is available to developers as an API. But instead the developers opted to use [this](https://github.com/IBM/ios-scanner) open source scanner. I at least partially blame this for poor scanning performance.
</details>

**Comment by timokoenig** [comment]
> Altering the schema to support this would be very simple.

This would mean to adjust all apps and rules of every member states so I would not consider this very simple.

> I think part of the issue of codes scanning badly is the specific QR code reader implemented in the CovPass apps.
> It seems to have some issues; especially when it is printed on a warped piece of paper such as the printout you get at a pharmacy. I have scanned far more complex barcodes with the iOS built-in QR code reader. As far as I am aware it is available to developers as an API. But instead the developers opted to use [this](https://github.com/IBM/ios-scanner) open source scanner. I at least partially blame this for poor scanning performance.

I'm not aware of any really big scanning issues with the iOS app except the usual problems with printouts. If you have any more information you are welcome to open up a ticket in the iOS repo. The used iOS library is just an UI wrapper around iOS internals so nothing special that affects scanning performance. Besides that it is known that there are scanning problems with Android and their SDKs which has been addressed in some other tickets.

**Comment by treysis** [comment]
> This would mean to adjust all apps and rules of every member states so I would not consider this very simple.
> 

Not necessarily. This could be a Germany-exclusive function. Similar to what Switzerland is doing with their Light-Certificate (which is only valid in Switzerland and can only be created for Switzerland-issued certificates).

**Comment by vaubaehn** [comment]
@Kamik423 
> This could be solved by making a request to an EU server that can "fuse" to certificates for the same user; creating a new signed certificate after validating the ones it gets. This combined vaccine+test QR code could then be shown in big screen phones.

Personally, I find this a nice and practical solution. The only downside would be, that test centers (or doctors) would need to additionally scan users' vaccination DCCs. This makes data privacy agreements much more complex, because test center/doctor/laboratory would either need to store the vaccination DCC before submitting test result together with the vaccination DCC to issuance portal for "fusion", and/or transferring the vaccination DCC over several instances (e.g., from test/center or doctor to laboratory before the laboratory transmits the result with vacc DCC in case of PCR test)... So, test centers would have more work and testees would have more to read (data privacy agreement) and sign... The original argument from EU legal department was, that the DCC should contain as least information as possible, and in case more information is needed, the presentation of multiple DCCs is better than integrating more information into one DCC...

> I think part of the issue of codes scanning badly is the specific QR code reader implemented in the CovPass apps.
> It seems to have some issues; especially when it is printed on a warped piece of paper such as the printout you get at a pharmacy. I have scanned far more complex barcodes with the iOS built-in QR code reader. As far as I am aware it is available to developers as an API. But instead the developers opted to use [this](https://github.com/IBM/ios-scanner) open source scanner. I at least partially blame this for poor scanning performance.

Yes, the never ending story about QR code readers... that hopefully will end soon (and only affected Android afaik):
The first implementation in CWA Android was the old ZXING library, that is not actively maintained anymore. This lib has problems with image processing, when single QR code blocks do not have an exact 1:1 proportion.
CovPass implemented a newer and still maintained library (boofCV scanner) for image processing, that is much more flexible, reliable and fast. But: due to bad stock Android implementations in Samsung devices, the lib crashed on newer models (folds...).

But CWA now adapted to boofCV, circumventing the crashes on Samsung fold devices, and I think CovPass/CovPassCheck will catch up with this - so QR code scanning should be solved for everyone soon.

**Comment by treysis** [comment]
> so QR code scanning should be solved for everyone soon.

Not until the stupid grey overlay is removed and/or QR codes can be presented in full brightness automatically.

**Comment by timokoenig** [comment]
> > This would mean to adjust all apps and rules of every member states so I would not consider this very simple.
> 
> Not necessarily. This could be a Germany-exclusive function. Similar to what Switzerland is doing with their Light-Certificate (which is only valid in Switzerland and can only be created for Switzerland-issued certificates).

Well that would be a different story. I personally would strongly advice against multi certificates because this comes with other issues that are not very easy to fix, eg. more scanning problems. The way Switzerland went with the light-certificate or what I called it half a year ago validation-certificate is IMO the right one. This would mitigate many of our problems and also increase data privacy for the user. I would even like to see this feature not only in Germany but in all member states

**Comment by vaubaehn** [comment]
> > so QR code scanning should be solved for everyone soon.
> 
> Not until the stupid grey overlay is removed

Can be... but isn't boofCV working well with this?

> and/or QR codes can be presented in full brightness automatically.

Call me wrong, but aren't CWA/CovPass-devs working on this?

**Comment by Ein-Tim** [comment]
> and/or QR codes can be presented in full brightness automatically.
>> Call me wrong, but aren't CWA/CovPass-devs working on this?

AFAIK this is already implemented under Android in CWA 2.16., see https://github.com/corona-warn-app/cwa-app-android/pull/4624

However, I can't reach this fullscreen mode in the Android CWA for certificates. Can you @vaubaehn?

**Comment by vaubaehn** [comment]
@Ein-Tim 

> However, I can't reach this fullscreen mode in the Android CWA for certificates. Can you @vaubaehn?

😆 I'm still on CWA 2.**6**.1 to be able to debug business rules with self-created certificates ❤️

**Comment by treysis** [comment]
ROFL. Again confirmed in my choice for using Swiss Covid with my German certificates. It's just so hassle-free.

**Comment by Kamik423** [comment]
On iOS the full brightness discussion was shot down in a hilarious way: https://github.com/Digitaler-Impfnachweis/covpass-ios/issues/23

Concerning fusion: The suggestion was to only create the fused QR code in the situation that the operator would have to scan multiple QR codes anyways (2G + Test). The app would still load in separate codes as it does right now but if there is a situation where multiple certificates are required it could request a server to fuse them. (Or just append both binary strings to one another as an easier but larger solution). This should however not only be restricted to Germany then as the entire point of the EUDCC spec is to be international.

**Comment by no-go** [comment]
I have something worry to have unintentionally trolled here :-(

Name, birthday and (because of "Janssen") vaccine name and date/age of vaccination as additional info in the CheckApp make unfortunately (data protection) sense to display them. Otherwise, these QR codes are just not distinguishable by the app from the "gatekeeper". A 2/2 instead of 3/3 or 1/1 (Janssen) can easily be printed on paper next to the QR code. Then anyone can claim it is the QR code of the vaccine and that (a 2nd print) of the booster and both times it is just Janssen (Johnson and Johnson).

But I am glad that something like 3/3 is at least coming soon. I hope the vaccine name will also be shown and this feature will not come 6 weeks after it has been decided that Janssen 1/1 is "unvaccinated" in bundesland XY or at the University of Düsseldorf (since 15. jan 2022). Actual I switch to the greenpass App, wich has this feratures and has a officialy apk download (but not in f-droid).

PEACE. I am not sure this issue is a discussion place for "everything wents nutty today".

**Comment by oliver-steinbrecher** [comment]
@vaubaehn  thanks for providing this good summary:

> I think the planned implementation is the best you can achieve with regard to a balance of usability, security (avoid abuse of fraud certificates), and data privacy. If I understand the planned implementation right,
> 
> * gatekeeper will chose a preset in CovPassCheck and select the admission rule according to legislation (e.g., 2G+...)
> * visitor will show the QR code from wallet app (CWA, CovPass)
> * gatekeeper scans the code. In case, a second QR code is needed - and only in this case - the gatekeeper will be informed to scan a 2nd QR code and of which type. User will switch to that 2nd QR code - and in current CWA this will be just a single fingertip (afair, this will also come to CovPass)
> * gatekeeper will get a signal for "pass" or "invalid" without revealing (unnecessary) details of health data of the visitor
> 
> That's it. So, it will ease the things significantly for visitors and for gatekeepers: clear information, still best possible data privacy, no discussions anymore about the booster status....

In addition there are discussion on EU level about new data structures to better support upcoming use cases. 

 Sometimes it looks easy to do a single change but generally it’s important that we working aligned with all participating countries. This ensures that the solution has an impact and helps us in our daily life’s. 

 We’re aware of the peoples needs and happy about these kind of discussions.

**Comment by r41d** [comment]
Given the political situation in Germany we would've needed this feature by November; for NRW, the largest federal state, 2G+ for many activities was enacted on December 28th, since then it's almost been a month.
I have never worked with an Android App written in Kotlin before but today I got curious and wanted to see if I could do a rudimentary Booster Checker and [this is what I ended up with](https://github.com/r41d/covpass-android/commit/167cac945fcfae458c32e68b629bc6c06891f170)
End result: https://imgur.com/a/eYtm5Q2
This is by no means a tested or polished version but I just wanted to see if a hacky version could be made within a day.
Feel free to experiment but please don't use this on foreign QR codes.

**Comment by Kamik423** [comment]
> In addition there are discussion on EU level about new data structures to better support upcoming use cases.

Are these being had publicly somewhere where we can follow what is going on? Like discussion transcripts or press releases?

**Comment by boecko** [comment]
When you want to get an impression, why this issue should have the highest priority of all. 
Just open the [Playstore-Review-Page](https://play.google.com/store/apps/details?id=de.rki.covpass.checkapp&showAllReviews=true) and read the last > 100 reviews.

Maybe someone should print them out on paper and give them to the Datenschutzbeauftragten of the RKI.

**Comment by iMonZ** [comment]
This is so silly. Why is it important how much boosters you have? Why isn’t it important when the last shot you had.

**Comment by mirabilos** [comment]
I’d argue for showing raw data and/or making the rules configurable ahead of time. For example, in our mini choir we’re going 2G+¹ even for boostered people, despite those being exempted from the legal requirement, as local policy. Same for local makerspace access, I’m told.

① actually 1G+ because everyone’s got theirs now, thankfully

**Comment by fm-sys** [comment]
Keep in mind that 2G+ also accepts people with recovery / 2 vaccinations without test in some federal states, if they are less than 90 days old.

NRW:
![2022-01-18-2](https://user-images.githubusercontent.com/64581222/150634880-fa061b21-8a11-4537-bdb3-2cf38300f19b.png)

Hesse:
![Screenshot_20220119_105024](https://user-images.githubusercontent.com/64581222/150634895-a09c840c-32f3-49a7-b2f1-1f11b28e039c.jpg)

**Comment by iMonZ** [comment]
> Keep in mind that 2G+ also accepts people with recovery / 2 vaccinations without test in some federal states, if they are less than 90 days old.
> 
> NRW: ![2022-01-18-2](https://user-images.githubusercontent.com/64581222/150634880-fa061b21-8a11-4537-bdb3-2cf38300f19b.png)
> 
> Hesse: ![Screenshot_20220119_105024](https://user-images.githubusercontent.com/64581222/150634895-a09c840c-32f3-49a7-b2f1-1f11b28e039c.jpg)

This actually makes more sense then what the other states are doing.

**Comment by r41d** [comment]
So, no progress on this?

**Comment by vaubaehn** [comment]
I'm no dev here, but there is (invisible) progress, which I know from previous discussions with devs. The problem is correct detection of the booster status. This is not a technical problem, but a political one: the federal states have very different definitions of what counts as a booster. For this, all information needs to be collected, and a new process for the wallet apps needs to be developed that accounts for the fast political changes and that is able to differ between the different federal states.
As CovPass(Check) and CWA are aligned in these basic functions, I can tell you that on CWA's side development has a good progress and will likely be adopted from CovPass soon.

**Comment by timokoenig** [comment]
@vaubaehn that is correct. @oliver-steinbrecher already mentioned earlier in their thread ago that the development has been completed and everything is on track. The notes for the upcoming release can be found here: https://github.com/Digitaler-Impfnachweis/covpass-android/discussions/172

**Comment by oliver-steinbrecher** [comment]
1.17 is going to be released next week. The development of 1.18 has been completed and is currently in our QA phase. :)

**Comment by vaubaehn** [comment]
@oliver-steinbrecher 
That's great news. To achieve the correct booster detection, Corona-Warn-App will implement a Common Covid Logic that extends respectively replaces CertLogic. I strongly suppose that Immunization Rules evaluated to determine admission state will be coded and deployed by TSI.
Will you have the same implementation or are you going a separate way?

**Comment by oliver-steinbrecher** [comment]
We’re evaluating the possibilities. In General CWA and CovPass teams are aligning alle activities related to DCC features. So stay tuned !

**Comment by vaubaehn** [comment]
@oliver-steinbrecher 
We (as in "the community")  will do some effort towards the publisher to get Immunization Rules as well as all other rules (like booster notification) into a public repository, so that we all may have an eye on them, and in best case, propose early changes in accordance with political changes, to speed up reliable updates.
It would be nice if you supported our effort in one of your next talks with the publisher.

(fyi: @Ein-Tim )

**Comment by Kamik423** [comment]
soool, does this mean it accepts boosters too? The update notes imply no. But reading this literally it means yes. This needs to be clarified in order to be useful for anyone.

![4DE2F912-3157-4F26-AF67-096037CD331A](https://user-images.githubusercontent.com/16278233/152153725-7e7dce68-55bb-4d31-8c59-7874e70dc0a2.jpeg)

**Comment by Kamik423** [comment]
Also why is there no 2G? There are situations where this is needed.

**Comment by Ein-Tim** [comment]
@Kamik423 

> does this mean it accepts boosters too?

This comes with the next version. Currently it still shows (even for boostered people), that a test is required.

**Comment by Kamik423** [comment]
Then it should really have a note there. This implementation will inevitably lead to lots of confusion. Let’s hope for 1.18 to be released soon.

**Comment by timokoenig** [comment]
@Kamik423 with the current release cycle the next release will be in 14 days. A little too late IMO

**Comment by r41d** [comment]
Just checked out the new release, it can neither check 2G, nor 2G+ via booster certificates, so it's still lagging behind.
The Apps from Austria and Switzerland have alle needed check modes since early December, just saying.

**Comment by Ein-Tim** [comment]
### TO-DOs

1. Detect booster vaccination and don't request a test certificate (**planned for 1.18**)
2. Integrate the rules of the different federal states (e.g. that a booster is only allowed to be 90 days old in some federal states, etc.).

**Comment by Karl-WE** [comment]
@oliver-steinbrecher as long there is no other PR for this I would like to propose the following.
Please check if this is consistent with your plans of 1.18:


1. Use day and time from the app polling ntp from ptb time. 

Reason: while the chance is minimal a manipulation or offset e. g. wrong summer/winter time from the mobile Provider could lead to wrong expiration of vaccination status. Now that the rules get more complicated and involve expiration based on number of days, allowance to boost after x days, etc correct time is essential. 



2. While the  CovPass / Check app is designed for use in different countries I would propose that the at least of not both apps check should offer the ability
- to select
- save
- or change
 a country, a german federal state or other country from a dropdown list in app settings. The new version should point out that before a check a valid country and or state has to be set up. 

This would reflect the needed flexibility needed when checking the credibility of the recovery and or vaccination status in different regions, due to unfortunate regional regulations. 


3. Covpass check app but also the regular app should provide an easy indicator about the overall status of the person at a glance. 
Again depending the selected country and / or federal state (#2) 

Boostered, vaccinated, recovered 
Due to Regulation of country, state from <effective date of regulation>, until < effective date until the state expires or degrades>


4. Selecting, saving, changing the entrance regulations in Covpass check app settings (2G, 2G plus, 3G etc.) 


5. Ultimately showing in the Covpass Check App wether according to selection of #4 and #2 a further (quick) test has to be Provided for entrance allowance. 


6. For both Apps a clear indicator, in concordance with #2 when a person has a degraded or invalid state

Usecase: this allows a self control next to the check app to avoid missing a new effective regulation that affects the state. 
e. g. J&J, 


If I missed anything please feel free to add. 
It's important that we try to catch all the complexity brought into that, because as others noted the ability to check the state is crucial for the usefulness of the apps. 

Hope this list can be checked again the 1.18 plans and helps to make this important release most successful. 
As others said, I understand the dependencies for such changes but time isn't in our favor. 

Currently there is a high risk of wrong entrance allowance.

Thanks for your time and feedback!

**Comment by Ein-Tim** [comment]
@Karl-WE 

> While the CovPass / Check app is designed for use in different countries I would propose that the at least of not both apps check should offer the ability to
- select
- save
- change
a country, a german federal state or other country from a dropdown list in app settings.

Do you have an overview over the different rules in the different federal states?
Again, I don't know what leads you to the conclusion that the CovPassCheck-App is intended to be used outside of Germany.

**Comment by Karl-WE** [comment]
Hello @Ein-Tim personally I don't have a list of all regulations per country / federal state. This could be either collected in a PR for future updates by the community or become the responsibility of the owners.

I have some where found a list in which EU countries outside of Germany the apps are supported, but unfortunately I didn't bookmark this reference. It was missing some bigger countries like Spain but included smaller countries like Andorra.

**Comment by Ein-Tim** [comment]
@Karl-WE 

> personally I don't have a list of all regulations per country / federal state. This could be either collected in a PR for future updates by the community or become the responsibility of the owners.

It's okay that you don't have such a list, the problem is that even BMG/RKI seem to not have such a list. 
Therefore, I'm also thrilled to see how the implementation in 1.18 will be.

> I have some where found a list in which EU countries outside of Germany the apps are supported, but unfortunately I didn't bookmark this reference. It was missing some bigger countries like Spain but included smaller countries like Andorra.

Yes, this is where the apps can be downloaded, however, Spain and all other EU countries have their own wallet & check apps. So CovPasCheck doesn't have to include their rules.

**Comment by Karl-WE** [comment]
Sounds like a typical case of german  regulations.

**Comment by sleeksorrow** [comment]
I don't think that CovPassCheck should know and evaluate all regulations from all regions and locations. I assume it's impossible to keep up with that.

But as of now its still not possible to gain all information from the output that is needed for a decision in 2G+ scenarios to decide if an additional test is needed or not. This is the most important thing: Show all needed info. Like "Booster #1 XX days ago" or maybe just "Immunisation 3/3 XX days ago" or similar, which _at the moment_ means the same as "fully boostered". It can be made prettier afterwards.

Of course the date of last immunisation is a sensitive information, but as government has decided that this value is part of the rules, it's obviously okay to share.

**Comment by Karl-WE** [comment]
Thank you for your opinion @sleeksorrow. Looking forward for the implementation in 1.18. Especially for the CovPass Check app. Without doubt it is confusing and from a programmatical point of view the solution should be sustainable.

I do not expect a near end of the situation and more boosters and / or more dependencies, this should be considered in the code. That's why I drafted a wide range solution from the logical point of view. As the app is owned by the RKI they **should** have all regulations that are valid for Germany, or they need to establish a political or business process to assure they receive the information, as @Ein-Tim mentioned they do not publish it at the moment. 

Currently, even though offtopic for the PR, the discussion about the recovered state and the duration shows that the political decisions favor a central publication for Germany of the regulations, but this is a heated debate, and is currently only tied to the recovery, not other aspects, afaik.

**Comment by Kamik423** [comment]
@sleeksorrow I disagree. The point of the app is to make checking regulations easier. This entails not juggling a bunch of numbers in your head to match against. Its entire purpose is it to show a big green check mark if a certificate is valid for entry and a big red x if it is not.

**Comment by mirabilos** [comment]
Hans dixit:

***@***.*** I disagree. The point of the app is to make checking
***@***.*** easier. This entails not juggling a bunch of numbers in
***@***.*** head to match against. Its entire purpose is it to show a big
***@***.*** check mark if a certificate is valid for entry and a big red x
***@***.*** it is not.

But this makes checking regulations harder or impossible
because regulations change, or there may be site-local,
venue-local, usecase-local etc. requirements (by law or
by site/venue policy).

**Comment by sleeksorrow** [comment]
I'd appreciate if the app can make checking easier and I approve any steps towards this even if I assume as well that this is technically impossible. But before we can make checking easier, we first need to make checking _possible_ which it is not as of now without displaying the age of latest immunization.

**Comment by Karl-WE** [comment]
all regulations are logical, so I suspect they can be coded. Depending on the structure it should be flexible enough to adjust the result to a number of combinations. I hope that the developers will also comment the ideas here. It's too important to get things wrong. Clock is ticking.

**Comment by sleeksorrow** [comment]
> all regulations are logical

I need to know all regulations first before I can decide on that statement. Please list them all in complete, for all german states, with all their requirements and all their exceptions. I'm desperately searching for such a list since May 2020.

> so I suspect they can be coded

Yes, I suspect that they can be coded, too. That's why I disagree and they should be second objective at best.

**Comment by Ein-Tim** [comment]
@sleeksorrow

> I need to know all regulations first before I can decide on that statement. Please list them all in complete, for all german states, with all their requirements and all their exceptions. I'm desperately searching for such a list since May 2020.

This is really the problem here, it seems like neither the RKI, nor the BMG have a public list over all the different regulations in the federal states. The development teams of the apps have to search the information by themselves and make sure they always update the information.

**Comment by Kamik423** [comment]
As stated in a post by me above, I think the best way would be to include a UI like this:

```
Valid checks:
[ ] Antigen Test
[ ] PCR Test
[ ] Vaccination
[X] Booster
[X] Vaccination + Test
```

There should be preset buttons (2G, 3G, 2G+) but it should have local concepts for what constitutes those (I.e. if 2G+ is fulfilled by a booster) and what counts as a booster (I.e. 2/1 or only 3/3).

I think this is the **necessary** thing to do, as this can cover all options that will be needed while remaining intuitive, accessible and flexible.

**Comment by sleeksorrow** [comment]
@Kamik423 I fully agree to keep it simple and let the checking person evaluate. I just want to mention that your proposal does not cover all options, because in 2G+ situations in some (all? dunno) federal states a full vaccination without booster can be enough without the additional test, if the date of immunization is less than 90 days. That's why the app needs to show the days since last immunization.

**Comment by Jo-Achim** [comment]
No question - of course there is a basic logic that can or could also be programmed. But the many, partly local, exceptions, their half-lives, etc. can quickly undermine a basic logic or make it significantly more difficult or even impossible to program it.

A good example is the shortening of the convalescent status to 3 months, introduced overnight - which is now being overturned by the courts. Under such framework conditions - and that was just a very simple example - I don't want to be a programmer / person in charge!
Unfortunately, neither the causes nor the consequences are seen correctly everywhere.
In this respect, comparisons with (foreign) apps are correct and important, but not always helpful because the structures there are perhaps a little more stringent. 

Nevertheless, time is ticking. Also because increasing loosening is at least being discussed, which wants to reduce or even end access restrictions.
However, I am afraid that we will not be lucky enough not to experience access restrictions again in the coming autumn / winter. Seen in this way, we may then experience the same discussion again.

**Comment by Karl-WE** [comment]
based on the information of Android the app is published by the RKI so ultimately they should have the responsibility about the reliability of checks @Ein-Tim. We are note discussing an entire open source application which is maintained by people in their free time, do we?

**Comment by Dale81** [comment]
@Karl-WE absolutely true, it not only a free-time project. Even if we do not know what is going on behind the scenes, I am also wondering why the app is not reliable at the moment. I know I am repeating this wherever I post something, but for me the currently "classic" example is the one-dose Johnson & Johnson vacc which is still represented valid as "fully vaccinated" - which is not true anymore. I don't have a recovery certificate at hand which would be over 90 days but still under 180 to check this, but I assume it would still be valid even the 90 days rule came in.

Can someone explain what the business rules are ment for as they are show in the app to be updated every now and then. I thought, these rules may include exactly these cases which could be updated ad hoc without a real app update via stores.

I really would like to understand this logic.

**Comment by boecko** [comment]
This is by no means a free-time project [CovPass wird fünfmal so teuer wie geplant](https://www.wiwo.de/technologie/digitale-welt/digitaler-impfpass-covpass-wird-fuenfmal-so-teuer-wie-geplant/27788324.html)

> Knapp 15,4 Millionen Euro kosten der digitale Impfnachweis CovPass und die damit verbundenen Clouddienste das Gesundheitsministerium und damit deutlich mehr als ursprünglich vorgesehen.

**Comment by dadaphl** [comment]
I think the 3g,2g+ check feature is broken and should better be removed entirely at this point.

For example, is no SINGLE certificate that can combine vaccinated-recovered-vaccinated status. This certificate simply can not be created in the pharmacies due to fact that the certificates are given by the EU. A certificate for the second vaccine will simply omit the information about the recovery. As the app currently only has one active certificate and checks are performed only on a single cert, it will show "Certificate does not fulfill all conditions" when 14 days haven't passed since the second shot.
In many states (Berlin for example) this constellation qualifies as "Boostered" and there is no 14 days waiting time required.

I think it would be better if the CovPassCheck App simply shows if the certificate is valid in general and not in regard to wrongly interpreted rules (that keep changing faster than the release cycle of this app anyway). 

It creates a lot of confusion and distrust among personal that is using the app to check validity. Imagine, they are dealing with all those edge cases several times a day. Many don't seem to take "Invalid Certificate" serious anymore. This is my personal observation. I have been chasing the reasons and responsibilities behind this edge case now since a week as I am one of the affected.

The RKI doesn't make the rules, it just releases guidelines. The states implement rules inspired by the RKI guidelines. The wording in those rules (Verordnungen) is ambiguous and fuzzy. If you call the authorities (Gesundheitsamt, Senatskanzlei, Corona Hotline) in Berlin 5 times you get 5 different interpretations on their own rules.  I think the developers can not expect to be given an "up to date" list of the regulations in the different states because it seems there is no authority that is responsible for such information.

Due to the complex nature of the problem, it is understandable that the App can not keep up. Please redesign the features in a way that it makes the users (personal that checks 3g/2g+) feel responsible to make their own judgement of the situation. At the moment, the app is misleading and wrongfully denying entry to perfectly immunized people.

**Comment by treysis** [comment]
The way it's currently going is just vain endeavor. Always too late, still too confusing. I was trying to keep up with it but even I am now lost as what constitutes what. A solution could be to have the rule set separately on the server, so not everyone has to wait for a new app release.

PS: Anyways, there are still fake mickey mouse certificates going around from over 3 months ago. So what's all the fuzz. This project is lost effort.

**Comment by Jo-Achim** [comment]
@dadaphl wrote:
> I think the 3g,2g+ check feature is broken and should better be removed entirely at this point.
> ...
> It creates a lot of confusion and distrust among personal that is using the app to check validity. Imagine, they are dealing with all those edge cases several times a day. 
> ...

In fact, today was the first time I experienced that a valid 'booster certificate' (my wife's) shown with CovPass (v 1.17.0) was supposedly invalid.
This is surprising because my vaccination certificate (we were vaccinated on the same dates with the same vaccines) was valid. My wife has a paper printout of her DCC for such cases - apparently (tested with the same smartphone) that was also displayed as invalid.

Unfortunately, there was no way of clarifying the circumstances, but the readable text "3/3" in the paper DCC led to admission. Since I assume that testing was done with CovPassCheck, I suspect either an operator error or an old version of the app.
Are such possible 'operator errors' known?

Back home I checked the supposedly invalid certificate with a second Android smartphone (CovPassCheck v1.17.0) - and it is displayed as valid under "Check 3G" (and under "Check 2G+" as "2G proof valid").


Conclusion: **I still don't think that the checks mentioned above are broken or that they are superfluous.**
As already mentioned, they just come a little late against the background of the opening discussion that is beginning - _at least for the current wave of infections_.

**Comment by dadaphl** [comment]
> Conclusion: **I still don't think that the checks mentioned above are broken or that they are superfluous.**
are you refering to the checkes mentioned in your own text or mine? It seems like that the operator made a mistake or the app was not up to date in your case. In my case however the apps are up to date and I can assure you there is no operator error going on. The checks are broken in the CovPassCheck app.

**Comment by Jo-Achim** [comment]
@dadaphl: 
Your question is easy to answer. I was essentially referring to
- the quote from your post,
- my experience yesterday,
- their unclear circumstances and
- my question is whether the problem described in this way is known.

In this respect I think it would be helpful if you open your own bug report and name all the circumstances you are aware of. So which smartphone was used, which operating system, which app version, ... was scanned from paper or from an display on another smartphone, how often the problem occurred / occurs etc.
These statements would surely help the app developers to find and eliminate any problems that may arise.

Why / what is the background to this? I remember several problems around certificate scanning. Be it that the smartphone libraries used had difficulties, be it that cell phone cameras didn't always want to focus or something similar.
To what extent these problems only applied to CovPass / CovPassCheck or the CWA, I don't know much, but the developers will know.

Therefore, from my point of view, my conclusion remains.

**Comment by Jo-Achim** [comment]
@Ein-Tim,

yes, the result of a 2G+ check is now displayed as shown.

With that I will close this point.
Although the text "Scan the necessary documents for 2G+ one by one." could be adjusted according to the switch position "No test necessary for booster vaccinations". Because if that switch is 'on' then there's no more evidence to scan for 2G+, right?
